### PR TITLE
Add ability to respond to a message reaction being added or remove 

### PR DESCRIPTION
**Pull request by ArcticSnowman**
## Proposed change

I worked on a slack chat bot that was based on https://github.com/ArcticSnowman/go-slackbot which I added the ability to respond to emojis (reaction).

So this is my first attempt to add reaction support to this bot framework.  Hope this helps with #45 

## Types of changes

What types of changes is this pull request introducing to flottbot? _Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

You can fill this out after creating your PR. _Put an `x` in the boxes that apply_

- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by wass3r — CHANGES_REQUESTED** [review]
thanks for the contribution! with a couple of caveats it works as advertised (see comments).

some further tweaks that would be interesting to explore are (and this definitely doesn't have to be in this PR):
1) separate the action (added/removed) so you can target either
1) check the target thread to see whether we already replied and, at least by default, don't post again

File: remote/slack/helper.go [hunk 1]
```diff
@@ -445,6 +445,27 @@ func populateMessage(message models.Message, msgType models.MessageType, channel
...
+// populateMessage - populates the 'Message' object to be passed on for processing/sending.
+func populateReaction(message models.Message, msgType models.MessageType, channel, action, reaction, timeStamp, link string, user *slack.User, bot *models.Bot) models.Message {
+	switch msgType {
+	case models.MsgTypeDirect, models.MsgTypeChannel, models.MsgTypePrivateChannel:
+		message.ReactionAction = action
+		message.Reaction = reaction
+		message.Input = ""
+		message.Output = ""
+		message.Timestamp = timeStamp
+		message.SourceLink = link
```
[the remark this answers is not in the document]
Comment (ArcticSnowman) [reply remote/slack/helper.go:457]: Fixed

File: config-example/rules/reaction.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+name: plus-one-reaction
+active: true
+reaction_match: ":+1:"
```
[the remark this answers is not in the document]
Comment (ArcticSnowman) [reply config-example/rules/reaction.yaml]: I've updated the example and fixed some issue I found.  I decided that we need to treat addition and removals of reactions are separate in Rule and Message model so action can be triggered for one or other or both.

File: Makefile [hunk 3]
```diff
@@ -63,7 +63,8 @@ test-coverage:
 clean: validate
 	@echo "Running $@ tasks"
 	-rm -v ./flottbot*
-	-rm -v ./debug
+
+##	-rm -v ./debug #Not created
```
Comment (wass3r) [inline Makefile:67]: let's undo this since it's unrelated to the feature

File: core/matcher.go [hunk 4]
```diff
@@ -63,25 +63,62 @@ RuleSearch:
...
+	} else if rule.ReactionsAdded != "" {
+		log.Debug().Msgf("ReactionAdded Rule %s", rule.Name)
+
+		messageReaction := message.ReactionAdded
+		processedInput, hit = utils.Match(rule.ReactionsAdded, messageReaction, false)
+	} else if rule.ReactionsRemoved != "" {
+		log.Debug().Msgf("ReactionRemoved Rule %s", rule.Name)
+
+		messageReaction := message.ReactionRemoved
+		processedInput, hit = utils.Match(rule.ReactionsRemoved, messageReaction, false)
```
Comment (wass3r) [inline core/matcher.go:91]: let's remove the debug log messages in this section. it basically prints an inventory of all rules every time a message gets processed. probably not a bad idea to do on startup where we already print the number of rules that you have setup (but not necessary to add in this PR).

File: core/matcher.go [hunk 5]
```diff
@@ -231,7 +270,7 @@ func isValidHitChatRule(message *models.Message, rule models.Rule, processedInpu
 	}
 
 	// If this wasn't a 'hear' rule, handle the args
-	if rule.Hear == "" {
+	if rule.Hear == "" && rule.ReactionsAdded == "" {
```
Comment (wass3r) [inline core/matcher.go:273]:
```suggestion
	if rule.Hear == "" && rule.ReactionsAdded == "" && rule.ReactionsRemoved == "" {
```

**Review by wass3r — CHANGES_REQUESTED** [review]
works as advertised. thanks for the tweaks. couple of small things and we're good i think.

i noticed you used plural `reactions_added` vs `reaction_added`. was the thought to be able to support match multiple reactions in one rule at some point? if so, i think i like that. some workspaces have multiple emojis that are similar, eg. `:question:` and `:grey_question:`. nothing that has to be fleshed out in this PR though. just curious.

thank you!

File: core/matcher.go (hunk 4 above)
Comment (ArcticSnowman) [reply core/matcher.go:91]: Removed

**Review by ArcticSnowman — COMMENTED** [review]
Should
```
	if rule.Hear == "" && rule.ReactionsAdded == "" && rule.ReactionsRemoved == "" {
```
not be better as 
```
	if rule.Respond != "" {
```

**Comment by wass3r** [comment]
> Should
> 
> ```
> 	if rule.Hear == "" && rule.ReactionsAdded == "" && rule.ReactionsRemoved == "" {
> ```
> 
> not be better as
> 
> ```
> 	if rule.Respond != "" {
> ```

makes a lot of sense, i like it! let's do that.

**Comment by ArcticSnowman** [comment]
@wass3r  - I went with plurals as the use case in the other bot library I use, does take multiple emoji triggers.

For this framework, the user would need to craft a regexp string to match multiple.

For example:

`(plus_one|plusone|plus1)`

**Comment by ArcticSnowman** [comment]
PR for docs https://github.com/target/flottbot-docs/pull/60

File: core/matcher.go [hunk 6]
```diff
@@ -209,6 +238,8 @@ func handleNoMatch(outputMsgs chan<- models.Message, message models.Message, hit
 }
 
 // isValidHitChatRule does additional checks on a successfully hit rule that came from the chat or CLI service.
+//
+//nolint:gocyclo // refactor
```
Comment (wass3r) [inline core/matcher.go:242]: looks like this addition is making the build and validate steps fail

**Comment by wass3r** [comment]
for the failing test (https://github.com/target/flottbot/actions/runs/11521113211/job/32081108474?pr=520#step:5:252), it looks like the improvement you made exposed a flaw in a testcase. in this block https://github.com/target/flottbot/blob/dc4cfeb2bc5ca489868bba23f9e12261ba70c132/core/matcher_test.go#L647-L653 we will need to add `testRuleNeedArg.Respond = "foo"` (for example)

**Comment by ArcticSnowman** [comment]
@wass3r - Yeah those unit tests need an explicit Hear/Respond added.   I've clean up the test and added a check to the function that requires one of Hear, Respond, ReactionsAdded, or ReactionsRemoved. to have a value to be valid.

**Comment by wass3r** [comment]
i think we're one commit away.. linter is (harshly) complaining about this line https://github.com/target/flottbot/blob/63542e733452e64d6d6c3063a6d74960ea900868/core/matcher.go#L244 . happy to merge after that tiny edit. thanks for all the work!

**Comment by ArcticSnowman** [comment]
Fixed to keep the linter happy...

On the topic of a rule checking if it's replied to the reaction already, the script/code handling the reaction would need to either
- Be able to call the Slack (or other service) API to interrogate if the bot had responded 
- Keep some stateful data to log the TimeStamp of the message reacted to.

**Comment by wass3r** [comment]
yup, i was thinking about utilizing https://api.slack.com/methods/conversations.replies . trying to shy away from adding the complexity of state management for users/admins at this time :)

**Review by wass3r — APPROVED** [review]
thank you!

**Review by wass3rw3rk — APPROVED** [review]

**Pull request merged**